analyzers/src/SonarAnalyzer.CSharp/Rules/ExpectedExceptionAttributeShouldNotBeUsed.cs

```diff
@@ -18,18 +18,51 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.Common.Walkers;
+
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     public sealed class ExpectedExceptionAttributeShouldNotBeUsed : ExpectedExceptionAttributeShouldNotBeUsedBase<SyntaxKind>
     {
         protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
 
-        protected override bool HasMultiLineBody(SyntaxNode syntax)
+        protected override bool HasMultiLineBody(SyntaxNode node)
         {
-            var declaration = (MethodDeclarationSyntax)syntax;
+            var declaration = (MethodDeclarationSyntax)node;
             return declaration.ExpressionBody is null
                 && declaration.Body?.Statements.Count > 1;
         }
+
+        protected override bool AssertInCatchFinallyBlock(SyntaxNode node)
+        {
+            var walker = new CatchFinallyAssertion();
+            foreach (var x in node.DescendantNodes(_ => true).Where(x => x.IsAnyKind(SyntaxKind.CatchClause, SyntaxKind.FinallyClause)))
+            {
+                walker.SafeVisit(x);
+            }
+            return walker.HasAssertion;
+        }
+
+        private sealed class CatchFinallyAssertion : SafeCSharpSyntaxWalker
+        {
+            public bool HasAssertion { get; set; }
+
+            public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+            {
+                HasAssertion = node.Expression
+                    .ToString()
+                    .SplitCamelCaseToWords()
+                    .Intersect(UnitTestHelper.KnownAssertionMethodParts)
+                    .Any();
+
+                if (HasAssertion)
+                {
+                    return;
+                }
+
+                base.VisitInvocationExpression(node);
```

Comment:
Is this needed? What would happen in the base version of the method?

Reply:
From my understanding, you must call to continue the visit; if you don't, it will stop.
So it would not find the assertion in this case for instance:

```cs
finally
{
  Console.Writeline("Something");
  Assert.True(true);
}
```

Reply:
I checked the code. The method in the base class just invokes a method called a virtual method `DefaultVisit`. This method is empty and it's not overwritten.
When I removed the call to the base method every test remained green. I think it's safe to remove.